### PR TITLE
Add usbserial kernel modules

### PR DESCRIPTION
**Pull request by fredrike**
_Motivation:_  Synology DSM is lacking usb-serial kernel support, this fixes that.

### Checklist
- [x] Build rule `all-supported` completed successfully
- [ ] Package upgrade completed successfully
- [x] New installation of package completed successfully

### CODE REVIEW

File: cross/ser2net/Makefile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+PKG_DIR = $(PKG_NAME)-$(PKG_VERS)
+PKG_DIR = $(PKG_NAME)-$(PKG_VERS)
+PKG_DIR = $(PKG_NAME)-$(PKG_VERS)
+PKG_DIR = $(PKG_NAME)-$(PKG_VERS)
+PKG_DIR = $(PKG_NAME)-$(PKG_VERS)
+PKG_NAME = ser2net
```
Comment (hgy59) [inline cross/ser2net/Makefile]: I looks like this file is not related to this PR

Comment (fredrike) [reply cross/ser2net/Makefile]: You are absolutely correct, thanks for spotting this.

**Comment by fredrike** [comment]
@hgy59 as this is kernel modules the build target must be the same as the running kernel, what is the reasoning for not having say `braswell-6.22`? I don't want to spend hours addi g the latest tool chains and kernels to realize that there is a policy against minor versions (or so).

**Comment by hgy59** [comment]
@fredrike I am not sure whether a build for kernel version 4.4.15 is compatible with 4.4.59 (or 3.10.102 with 3.10.105) but you are right, there are some kernel updates from DSM 6.1 to DMS 6.2.2

- 2.6.32 no update since DSM 3.2 [QorIQ and PPC853x]
- 3.2.40 no update since DSM 5.2 (updated from 3.2.11 and 3.2.30 with DSM 4.2 to 3.2.40 with DSM 4.3)
- 3.10.105 starting at 3.10.35 with DSM 5.2, updated to 3.10.77 with DSM 6.0, updated to 3.10.102 with DSM 6.1, updated to 3.10.105 with DSM 6.2.2
- 4.4.59 starting at 4.4.15 with DSM 6.1, updated to 4.4.59 with DSM 6.2

So only the 3.10.x and 4.4.x kernels got updates within the range of DSM 5.2 to 6.2.2
(and DSM 7 is already announced)

<h3>We could build the following [kernel-version]</h3>

Regarding the **all-supported** archs (ppc853x-5.2; others with TC 6.1) 

- 88f6281-6.1 [2.6.32]
- aarch64-6.1 (not a target, use rtd1296 and armada37xx)
  - rtd1296-6.1  [4.4.15]
  - armada37xx (not available for TC 6.1)
- alpine-6.1 [3.10.102]
- armada38x-6.1  [3.10.102]
- armada370-6.1  [3.2.40]
- armada375-6.1  [3.2.40]
- armadaxp-6.1  [3.2.40]
- comcerto2k-6.1  [3.2.40]
- evansport-6.1  [3.2.40]
- hi3535-6.1  [3.4.35]
- monaco-6.1  [3.10.102]
- ppc853x-5.2 [2.6.32]
- qoriq-6.1 [2.6.32]

and for x64 archs:
- apollolake-6.1  [4.4.15]
- avoton-6.1   [3.10.102]
- braswell-6.1   [3.10.102]
- broadwell-6.1   [3.10.102]
- broadwellnk-6.1  (not available for TC 6.1)
- bromolow-6.1   [3.10.102]
- cedarview-6.1   [3.10.102]
- denverton-6.1   [4.4.15]
- dockerx64-6.1  _(I would not recommend virtual environments)_
- grantley-6.1   [3.10.102]
- purley-6.1  (not available for TC 6.1)
- kvmx64-6.1  _(I would not recommend virtual environments)_

<h3>Additional archs for newer TC versions:</h3>

- rtd1296-6.2   [4.4.59]
- armada37xx-6.2.2   [4.4.59]
- alpine-6.2.2   [3.10.105]
- armada38x-6.2.2  [3.10.105]
- monaco-6.2.2   [3.10.105]
- apollolake-6.2  [4.4.59]
- avoton-6.2.2   [3.10.105]
- braswell-6.2.2   [3.10.105]
- broadwell-6.2.2   [3.10.105]
- broadwellnk-6.2   [4.4.59]
- bromolow-6.2.2   [3.10.105]
- cedarview-6.2.2   [3.10.105]
- denverton-6.2.2   [4.4.59]
- grantley-6.2.2   [3.10.105]
- purley-6.2.2   [4.4.59]

This results in 21 packages to build for all-supported archs and additional 15 packages for newer archs.
To support the SRM-Toolchains there would be another 5 packages
- northstarplus-1.1  [3.6.5]
- northstarplus-1.2  [3.6.11]
- ipq806x-1.1    [3.4.102]
- ipq806x-1.2    [4.4.60]
- dakota-1.2    [4.4.60]

The following archs are not supported yet on spksrc  (but included above)
- dakota
- purley

<h3>Proposal</h3>

I would propose first to build the packages of the current toolchain versions available in spksrc that are already on the latests kernel version.
- 88f6281-6.1 [2.6.32]
- **armada370-6.1  [3.2.40]**
- armada375-6.1  [3.2.40]
- armadaxp-6.1  [3.2.40]
- comcerto2k-6.1  [3.2.40]
- evansport-6.1  [3.2.40]
- hi3535-6.1  [3.4.35]
- **ppc853x-5.2 [2.6.32]**
- qoriq-6.1 [2.6.32]

In the mean time I could add the missing Toolchains to your branch to support all the latests kernels and we would build:
- rtd1296-6.2   [4.4.59]
- armada37xx-6.2.2   [4.4.59]
- alpine-6.2.2   [3.10.105]
- armada38x-6.2.2  [3.10.105]
- monaco-6.2.2   [3.10.105]
- **apollolake-6.2  [4.4.59]**
- avoton-6.2.2   [3.10.105]
- braswell-6.2.2   [3.10.105]
- broadwell-6.2.2   [3.10.105]
- broadwellnk-6.2   [4.4.59]
- bromolow-6.2.2   [3.10.105]
- cedarview-6.2.2   [3.10.105]
- denverton-6.2.2   [4.4.59]
- grantley-6.2.2   [3.10.105]
- purley-6.2.2   [4.4.59]
- northstarplus-1.2  [3.6.11]
- ipq806x-1.2    [4.4.60]
- dakota-1.2    [4.4.60]

_(I have hardware to test the archs in **bold**)_

**With this two steps we will support all the archs with the latest kernel versions.**
For other kernel versions the users will have to build the packages of their own.
With this scheme we will result in 27 instead of 41 packages to publish.

**Comment by ymartin59** [comment]
@hgy59 Great review.
As most (or even all, maybe) 6.1 architectures have received 6.2 updates, why not consider we would only provide that specific package for "latest" versions only?

**Comment by ymartin59** [comment]
@fredrike Many thanks for this contribution. There was a lack about these drivers, only available as binaries from jadahl.com until now...

**Comment by fredrike** [comment]
> @hgy59 Great review.
> As most (or even all, maybe) 6.1 architectures have received 6.2 updates, why not consider we would only provide that specific package for "latest" versions only?

I'm not really sure what you want me to fix from this review.

**Comment by hgy59** [comment]
Here are the latests toolchains and kernels.
Unfortunately for the new ARCH `purley` there is no kernel source available yet.
@ymartin59 is it a problem that with the new toolchains `make all-supported` does not build the TC 6.1, instead there will be TC 6.2 or TC 6.2.2?

**Comment by hgy59** [comment]
For an unknown reason `make all-supported` tries to build `broadwell-6.2` - that does not have toolchain and kernel Makefiles - instead of `broadwell-6.2.2`.
The other 28 packages are successfully compiled with `make all-supported`. Since  `all-supported` includes `dockerx64` and `kvmx64` there is a total of 29 packages.

_To build ppc arches I added the following code to `cross/libusb/Makefile`_
```
ifeq ($(findstring $(ARCH), $(PPC_ARCHES)),$(ARCH))
LIBUSB_AUTOGEN_PARAMS += --host=ppc-linux
endif
```

The first build takes some time, since there are large archives of kernel-sources to download.

_with `make kernel-digests` all 88 kernel sources referenced in `srcspk` are downloaded ending in nearly 70GB (most of the >1 GB sources are older kernels)_

File: cross/usb-serial/Makefile [hunk 2]
```diff
@@ -19,7 +19,7 @@ include ../../mk/spksrc.cross-cc.mk
 myCompile:
 	@mkdir -p $(WORK_DIR)/$(PKG_DIR)
 	@echo "obj-m += usbserial.mod.o" > $(WORK_DIR)/$(PKG_DIR)/Makefile
-	@echo "obj-m += usb-serial.o" > $(WORK_DIR)/$(PKG_DIR)/Makefile
+	@echo "obj-m += usb-serial.o" >> $(WORK_DIR)/$(PKG_DIR)/Makefile
```
Comment (hgy59) [inline cross/usb-serial/Makefile]: I did not verify whether it matters that usebserial.mod.o was missing in the generated Makefile

File: spk/usbserial/Makefile [hunk 3]
```diff
@@ -1,21 +1,23 @@
 SPK_NAME = UsbSerialDrivers
 SPK_VERS = 6
-SPK_REV = 7
+SPK_REV = 1
```
Comment (hgy59) [inline spk/usbserial/Makefile]: initial packages should start with (package-) revision 1

maybe for `SPK_VERS` we start at 1 too (or use 7 or 7.00 or 7.0.0 since usbutils is version 700)
or is there any other version of an included component available?

File: spk/usbserial/Makefile [hunk 4]
```diff
@@ -1,21 +1,23 @@
...
 DEPENDS = cross/usb-serial cross/usbutils cross/ftdi_sio cross/ch341 cross/cp210x cross/pl2303 cross/ti_usb_3410_5052
 
-REQUIRED_DSM = 6.1
+REQUIRED_DSM = 5.2
+# armarch64 is a group or archs (it should not build kernel modules, but all-supported does not exclude this)
+UNSUPPORTED_ARCHS = aarch64
 
-MAINTAINER = Jadahl
-DESCRIPTION = "USBSerialDrivers for DSM 6.2xx. This package include usbserial.ko ch341.ko cp210x.ko pl2303.ko ti_usb3410_5052.ko and ftdi_sio.ko. USB utils: Log in to your Synology with Telnet or ssh and type: lsusb -v or usb-devices. If you use my driver package, you might consider a donation with paypal. https://www.paypal.me/jumbotroll."
+MAINTAINER = SynoCommunity
```
Comment (hgy59) [inline spk/usbserial/Makefile]: IMO the maintainer is the one that is maintaining the synology spk and not the developer of the original source.
Feel free to use your github username.

File: spk/usbserial/Makefile [hunk 5]
```diff
@@ -1,21 +1,23 @@
...
+DESCRIPTION = "USBSerialDrivers includes usbserial.ko ch341.ko cp210x.ko pl2303.ko ti_usb3410_5052.ko and ftdi_sio.ko. USB utils: Log in to your Synology with Telnet or ssh and type: lsusb -v or usb-devices. If you use my driver package, you might consider a donation with paypal. https://www.paypal.me/jumbotroll."
 
 
 RELOAD_UI = yes
 STARTABLE = no
-DISPLAY_NAME = UsbSerialDrivers for DSM 6.2
+DISPLAY_NAME = UsbSerialDrivers
```
Comment (hgy59) [inline spk/usbserial/Makefile]: Since the package can be built for any DSM version (updated to REQUIRED_DSM = 5.2) and the TC-Version is part of a generated spk file I would like to remove the DSM version here and in DESCRIPTION.

**Comment by hgy59** [comment]
> 
> 
> @hgy59 Great review.
> As most (or even all, maybe) 6.1 architectures have received 6.2 updates, why not consider we would only provide that specific package for "latest" versions only?

My Proposal results in building the latest versions only.

**Comment by ymartin59** [comment]
@fredrike Are you "MAINTAINER = Jadahl"? Where these modules' sources for spksrc comes from?

**Comment by Jumbotroll** [comment]
This is partially forked from the synocommunity, but the USB modules have been updated by me but do not want any credit for this work, just use my work freely!

Regards
Jadahl

**Comment by fredrike** [comment]
@hgy59 Do you want to take over this code?

File: toolchains/syno-ppc824x-3.1/Makefile [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+TC_VERS = 3.1
+TC_FIRMWARE = 3.1-1594
+
+# Kernel 2.6.24
+#TC_DIST_NAME = gcc334_glibc233_ppc824x.tgz
+# Kernel 2.4.22
+TC_DIST_NAME = gcc334_glibc233_ppc.tgz
+
+TC_EXT = tgz
+#TC_DIST_SITE = https://sourceforge.net/projects/dsgpl/files/DSM%203.1%20Tool%20Chains/PowerPC%20824x%20Linux%202.6.24
```
Comment (ymartin59) [inline toolchains/syno-ppc824x-3.1/Makefile:13]: Please remove commented lines. Does DSM 3.1 on ppc824x really have two support kernel versions? If so, I would split into two distinct toolchains.

File: toolchains/syno-ppc824x-3.1/Makefile [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+TC_DIST_SITE = https://sourceforge.net/projects/dsgpl/files/DSM%203.1%20Tool%20Chains/PowerPC%20824x%20Linux%202.4.22
+
+TC_BASE_DIR = powerpc-linux
+TC_PREFIX = powerpc-linux
+TC_TARGET = powerpc-unknown-linux
+
+TC_CFLAGS = -I$(WORK_DIR)/$(TC_BASE_DIR)/$(TC_BASE_DIR)/include
+TC_CPPFLAGS = -I$(WORK_DIR)/$(TC_BASE_DIR)/$(TC_BASE_DIR)/include
+TC_CXXFLAGS = -I$(WORK_DIR)/$(TC_BASE_DIR)/$(TC_BASE_DIR)/include
+TC_LDFLAGS = -I$(WORK_DIR)/$(TC_BASE_DIR)/$(TC_BASE_DIR)/lib
```
Comment (ymartin59) [inline toolchains/syno-ppc824x-3.1/Makefile:23]: These lines are identical in most toolchains... I would set these values as defaults in `spksrc.tc.mk` if not already set in arch specific toolchain (like armv7 toolchains where additional options are required)

File: toolchains/syno-purley-6.2.2/Makefile [hunk 8]
```diff
@@ -16,7 +16,13 @@ TC_TARGET = x86_64-pc-linux-gnu
 TC_CFLAGS = -I$(WORK_DIR)/$(TC_BASE_DIR)/$(TC_BASE_DIR)/sys-root/usr/include
 TC_CPPFLAGS = -I$(WORK_DIR)/$(TC_BASE_DIR)/$(TC_BASE_DIR)/sys-root/usr/include
 TC_CXXFLAGS = -I$(WORK_DIR)/$(TC_BASE_DIR)/$(TC_BASE_DIR)/sys-root/usr/include
-TC_LIBDIR = $(TC_BASE_DIR)/sys-root/lib
-TC_LDFLAGS = -L$(WORK_DIR)/$(TC_BASE_DIR)/$(TC_LIBDIR)
+TC_LDFLAGS = -L$(WORK_DIR)/$(TC_BASE_DIR)/$(TC_BASE_DIR)/sys-root/lib
```
Comment (ymartin59) [inline toolchains/syno-purley-6.2.2/Makefile:19]: These lines are identical in most toolchains... I would set these values as defaults in `spksrc.tc.mk` if not already set in arch specific toolchain (like armv7 toolchains where additional options are required)

File: toolchains/syno-purley-6.2.2/Makefile [hunk 9]
```diff
@@ -16,7 +16,13 @@ TC_TARGET = x86_64-pc-linux-gnu
...
+
+FIX_TARGET = myFix
 
 include ../../mk/spksrc.tc.mk
+
+.PHONY: myFix
+myFix:
+	chmod -R u+w $(WORK_DIR)
+	@find $(WORK_DIR)/$(TC_BASE_DIR) -type f -name '*.la' -exec sed -i -e "s|^libdir=.*$$|libdir='$(WORK_DIR)/$(TC_BASE_DIR)/$(TC_BASE_DIR)/sys-root/lib'|" {} \;
```
Comment (ymartin59) [inline toolchains/syno-purley-6.2.2/Makefile:28]: This "myFix" target is copy-paste in many "recent" toolchains... I would move the target itself in `spksrc.tc-fix.mk` with a relevant "default target" (probably the recent version) the same way `spksrc.install.mk` provides `install_target` and `install_destdir_target` that can be used as `INSTALL_TARGET` value.

File: cross/usbutils/Makefile [hunk 10]
```diff
@@ -0,0 +1,17 @@
+PKG_NAME = usbutils
+PKG_VERS = 007
+PKG_EXT = tar.gz
+PKG_DIST_NAME = $(PKG_NAME)-$(PKG_VERS).$(PKG_EXT)
+PKG_DIST_SITE = http://ftp.iij.ad.jp/pub/linux/kernel/linux/utils/usb/usbutils/
```
Comment (ymartin59) [inline cross/usbutils/Makefile:5]: Please generate digest in `cross/usbutils/digests`

**Comment by ymartin59** [comment]
@hgy59 I think so many additional folders in "cross/" will make it difficult to navigate code for applications and libraries. If really relevant to have such "cross-cc" usage (see below), I would prefer to move these into a new folder "modules/".
I have found a large copy-paste pattern here. Instead of diverting "cross-cc" by replacing targets, what do you think about creating a dedicated generic Makefile in "mk/" to compile module list from a variable `KERNEL_MODULES`? That generic Makefile can be included directly from spk/usb-serial with list of modules to build and include in generated `PLIST` too.

File: toolchains/syno-purley-6.2.2/Makefile (hunk 8 above)
Comment (hgy59) [reply toolchains/syno-purley-6.2.2/Makefile:19]: I already started to devine a TC_LIBDIR and to adjust the FIX_TARGET in spksrc.tc.mk but moved it to a separate local work. I will try to merge that into this PR.

**Comment by hgy59** [comment]
> Please discard empty `digests` files.

missing digests files will generate a warning `===>  No digests file for ...`.
So as we have empty `PLIST` files in spk Makefile we should have empty digests files when no source is downloaded.
But this may be obsolete anyway with the proposed kernel modules build.

**Comment by th0ma7** [comment]
@ymartin59 moving the discussion here (from https://github.com/SynoCommunity/spksrc/pull/4399#issuecomment-770223152).

Quite an interesting PR.  I can certainly help out based on the recent rework of the kernel build functionality with some of my findings for `linuxtv` #4144. Also sub-versioning build (e.g. 6.x.x) publishing got fixed as well in another PR so really, ground work is now ready for this sort of PR.

First, it needs a major cleanup of all the now obsolete `kernel*` changes and migrate to using `KERNEL_REQUIRE`.  @fredrike are you still interested in pursuing this quest?  I can help you out to build the foundation for other similar work in the future.

Also, considering the amount of `cross/*` I wonder if we shouldn't create a `cross/kernel/*` or `spksrc/modules` or similar where linuxtv would fall into as well (food for thoughts I guess).

**Comment by th0ma7** [comment]
@hgy59 and @ymartin59 I've extracted the interesting bits and included them into my repo.
After playing with the code for a bit an idea for a module kernel framework came to light.
I'll try come back with something in a few days...

**Comment by fredrike** [comment]
>  @fredrike are you still interested in pursuing this quest? I can help you out to build the foundation for other similar work in the future.

Nah, I'm not super excited about spending more time on this.

**Comment by hgy59** [comment]
completed with #5746 and published as  synokernel-usbserial v1.1-8
but actually broken (see #6085)
